### PR TITLE
Starting point for Component Testing Utilities 

### PR DESCRIPTION
**Pull request by htdvisser**
#### Summary

This is the starting point for the component testing utilities we discussed.

Needed for https://github.com/TheThingsIndustries/lorawan-stack/pull/1744

#### Changes

- Added `pkg/component/test` utils (`pkg/util/test` would create import cycles)

#### Notes for Reviewers

- Main reviewer is @johanstokking; see `pkg/component/test`.
- All other changes are just search/replace.
- There is still alternative logic in the NS, because tests in the NS didn't follow the common pattern of setting up the component.
- This is just a starting point. As next steps we'll move other common use cases to the test utilities: setting `AllowInsecureForCredentials`, hooking up the cluster, ...

### CODE REVIEW

File: pkg/applicationserver/grpc_deviceregistry_test.go [hunk 1]
```diff
@@ -26,6 +26,7 @@ import (
 	. "go.thethings.network/lorawan-stack/pkg/applicationserver"
 	"go.thethings.network/lorawan-stack/pkg/auth/rights"
 	"go.thethings.network/lorawan-stack/pkg/component"
+	. "go.thethings.network/lorawan-stack/pkg/component/test"
```
Comment (rvolosatovs) [inline pkg/applicationserver/grpc_deviceregistry_test.go]: IDEM

File: pkg/applicationserver/applicationserver_test.go [hunk 2]
```diff
@@ -240,7 +241,7 @@ hardware_versions:
 	}()
 	time.Sleep(Timeout)
 
-	c := component.MustNew(test.GetLogger(t), &component.Config{
+	c := NewComponent(t, &component.Config{
```
Comment (rvolosatovs) [inline pkg/applicationserver/applicationserver_test.go]: Do not use `.` imports

**Comment by htdvisser** [comment]
@johanstokking see PR description:

> `pkg/util/test` would create import cycles

@rvolosatovs I think it's okay to dot-import test utilities in test files that use them and I don't think that using named imports here going to make the test code look any nicer. But since I really need to get this merged to move on, I changed it for you.

File: pkg/applicationserver/grpc_deviceregistry_test.go [hunk 3]
```diff
@@ -26,6 +26,7 @@ import (
 	. "go.thethings.network/lorawan-stack/pkg/applicationserver"
 	"go.thethings.network/lorawan-stack/pkg/auth/rights"
 	"go.thethings.network/lorawan-stack/pkg/component"
+	componenttest "go.thethings.network/lorawan-stack/pkg/component/test"
```
Comment (rvolosatovs) [inline pkg/applicationserver/grpc_deviceregistry_test.go]: TBH I think we should just name the package itself `componenttest`. I don't see a use case where the package would be imported as `test`

**Comment by rvolosatovs** [comment]
@htdvisser I also think that the functionality should be added to package `test`. What is the import cycle exactly? Is it just caused by some tests still not using the `_test` package?

**Comment by johanstokking** [comment]
> What is the import cycle exactly? Is it just caused by some tests still not using the `_test` package?

I guess that testing in `pkg/component` depends on `pkg/util/test` indirectly.

On another thought, I think it's better to keep testing as local as possible to packages. A repository-wide `pkg/util/test` with more than simple test utilities is asking for dependency trouble. This is a great example to keep local to `pkg/component`.

**Comment by htdvisser** [comment]
Go is complaining about _at least_ these cycles:

```
package go.thethings.network/lorawan-stack/pkg/auth (test)
        imports go.thethings.network/lorawan-stack/pkg/util/test
        imports go.thethings.network/lorawan-stack/pkg/component
        imports go.thethings.network/lorawan-stack/pkg/interop
        imports go.thethings.network/lorawan-stack/pkg/auth
import cycle not allowed in test
package go.thethings.network/lorawan-stack/pkg/auth/rights (test)
        imports go.thethings.network/lorawan-stack/pkg/util/test
        imports go.thethings.network/lorawan-stack/pkg/component
        imports go.thethings.network/lorawan-stack/pkg/auth/rights
import cycle not allowed in test
package go.thethings.network/lorawan-stack/pkg/component (test)
        imports go.thethings.network/lorawan-stack/pkg/util/test
        imports go.thethings.network/lorawan-stack/pkg/component
import cycle not allowed in test
package go.thethings.network/lorawan-stack/pkg/web (test)
        imports go.thethings.network/lorawan-stack/pkg/util/test
        imports go.thethings.network/lorawan-stack/pkg/component
        imports go.thethings.network/lorawan-stack/pkg/interop
        imports go.thethings.network/lorawan-stack/pkg/web
```

I don't think it's a good idea to rewrite all test packages to use `_test` for the sake fo being able to use a single, repo-wide `pkg/util/test` package without risking import cycles. Doing that would make it nearly impossible to do whitebox testing, and I think it's important to be able to do that.

I'm not a big fan of our current `pkg/util/test` anyway. I think it already (indirectly) imports way too many packages (that can therefore not do whitebox testing):

```
go.thethings.network/lorawan-stack/pkg/auth/cluster
go.thethings.network/lorawan-stack/pkg/cluster
go.thethings.network/lorawan-stack/pkg/config
go.thethings.network/lorawan-stack/pkg/crypto
go.thethings.network/lorawan-stack/pkg/crypto/cryptoutil
go.thethings.network/lorawan-stack/pkg/devicerepository
go.thethings.network/lorawan-stack/pkg/errorcontext
go.thethings.network/lorawan-stack/pkg/errors
go.thethings.network/lorawan-stack/pkg/events
go.thethings.network/lorawan-stack/pkg/events/fs
go.thethings.network/lorawan-stack/pkg/fetch
go.thethings.network/lorawan-stack/pkg/fillcontext
go.thethings.network/lorawan-stack/pkg/frequencyplans
go.thethings.network/lorawan-stack/pkg/gogoproto
go.thethings.network/lorawan-stack/pkg/i18n
go.thethings.network/lorawan-stack/pkg/jsonpb
go.thethings.network/lorawan-stack/pkg/log
go.thethings.network/lorawan-stack/pkg/metrics
go.thethings.network/lorawan-stack/pkg/random
go.thethings.network/lorawan-stack/pkg/redis
go.thethings.network/lorawan-stack/pkg/rpcclient
go.thethings.network/lorawan-stack/pkg/rpcmetadata
go.thethings.network/lorawan-stack/pkg/rpcmiddleware
go.thethings.network/lorawan-stack/pkg/rpcmiddleware/fillcontext
go.thethings.network/lorawan-stack/pkg/rpcmiddleware/hooks
go.thethings.network/lorawan-stack/pkg/rpcmiddleware/rpclog
go.thethings.network/lorawan-stack/pkg/rpcmiddleware/sentry
go.thethings.network/lorawan-stack/pkg/rpcmiddleware/validator
go.thethings.network/lorawan-stack/pkg/rpcserver
go.thethings.network/lorawan-stack/pkg/ttnpb
go.thethings.network/lorawan-stack/pkg/types
go.thethings.network/lorawan-stack/pkg/unique
go.thethings.network/lorawan-stack/pkg/util/randutil
go.thethings.network/lorawan-stack/pkg/version
```

So in the future I'd very much prefer to get rid of those rather than adding more, and from @johanstokking's comment, I guess he feels the same way.

**Pull request merged**